cmd/snap/cmd_run.go

```diff
@@ -979,3 +1017,132 @@ func (x *cmdRun) runSnapConfine(info *snap.Info, securityTag, snapApp, hook stri
+	// StartTransientUnit as "ssa(sv)a(sa(sv))". The signature can be
+	// decomposed as follows:
+	//
+	// Partial documentation, at the time of this writing, is available at
+	// https://www.freedesktop.org/wiki/Software/systemd/dbus/
+	//
+	// unitName string // name of the unit to start
+	// jobMode string  // corresponds to --job-mode= (see systemctl(1) manual page)
+	// properties []struct{
+	//   Name string
+	//   Value interface{}
+	// } // properties describe properties of the started unit
+	// auxUnits []struct {
+	//   Name string
+	//   Properties []struct{
+	//   	Name string
+	//   	Value interface{}
+	//	 }
+	// } // auxUnits describe any additional units to define.
+	type property struct {
+		Name  string
+		Value interface{}
+	}
+	type auxUnit struct {
+		Name  string
+		Props []property
+	}
+
+	// We ask the kernel for a random UUID. We need one because each transient
+	// scope needs a unique name. The unique name is comprosed of said UUID and
+	// the snap security tag.
+	uuid, err := randomUUID()
+	if err != nil {
+		return err
+	}
+	// Instead of enforcing uniqueness we could join an existing scope but this has some limitations:
+	// - the originally started scope must be marked as a delegate, with all consequences.
+	// - the method AttachProcessesToUnit is unavailable on Ubuntu 16.04
+	//
+	// TODO: invert the arguments so that security tag comes first, UUID
+	// second. This will result in nicer scope names but needs to be
+	// coordinated with code in sandbox/cgroup and with the spread tests
+	// exercising that.
+	unitName := fmt.Sprintf("snap.%s.%s.scope", uuid, strings.TrimPrefix(securityTag, "snap."))
+	mode := "fail"
+	properties := []property{{"PIDs", []uint{uint(os.Getpid())}}}
+	aux := []auxUnit(nil)
+	systemd := conn.Object("org.freedesktop.systemd1", "/org/freedesktop/systemd1")
+	call := systemd.Call(
+		"org.freedesktop.systemd1.Manager.StartTransientUnit",
+		0, /* call flags */
+		unitName,
+		mode,
+		properties,
+		aux,
+	)
+	var job dbus.ObjectPath
+	if err := call.Store(&job); err != nil {
+		if dbusErr, ok := err.(dbus.Error); ok {
+			switch dbusErr.Name {
+			case "org.freedesktop.DBus.Error.UnknownMethod":
+				// The DBus API is not supported on this system. This can happen on
+				// very old versions of Systemd, for instance on Ubuntu 14.04.
+				logger.Debugf("cannot create transient scope on this system: %s", err)
+				return nil
+			case "org.freedesktop.systemd1.UnitExists":
+				// Starting a scope with a name that already exists is an
+				// error. Normally this should never happen.
+				return fmt.Errorf("cannot create transient scope: scope %q clashed: %s", unitName, err)
+			case "org.freedesktop.DBus.Error.Spawn.ChildExited":
+				// The system does not have a DBus session bus.
+				// Consume the error and continue with extra checks below.
+				logger.Debugf("session bus is not available, consider installing dbus-user-session")
+				err = nil
```

Comment:
I agree with the sentiment here but I wonder if we want to adjust our the distro packaging to ensure dbus-user-session is installed on those distros that don't have it installed. Also, is this error message distro-agnostic enough? Basically I'm wondering if there is a way for us to have refresh app awareness on systems more easily than the admin having to add something else (eg, consider a remote tmux session in the cloud that has a long-running snap command going).

This is not a blocker for this PR.

Reply:
I think that this packaging weirdness is really specific to Debian. Everyone else provides session bus by default